test/gateway/eip20_gateway/revert_stake.js

```diff
@@ -52,9 +66,289 @@ contract('EIP20Gateway.revertStake()', function (accounts) {
+    );
+
+    assert.strictEqual(
+      returnedValues.staker_,
+      stakeMessage.staker,
+      `Returned staker ${returnedValues.staker_} value is different from expected ${stakeMessage.staker} `,
+    );
+    assert.strictEqual(
+      returnedValues.stakerNonce_.eq(stakeMessage.stakerNonce),
+      true,
+      `Returned staker nonce ${returnedValues.stakerNonce_.toNumber(10)}`
+      + ` value is different from expected ${stakeMessage.stakerNonce.toNumber(10)} `,
+    );
+    assert.strictEqual(
+      returnedValues.amount_.eq(stakeRequest.stakeAmount),
+      true,
+      `Returned amount ${returnedValues.amount_.toNumber(10)}`
+      + ` value is different from expected ${stakeRequest.stakeAmount.toNumber(10)} `,
+    );
+
+  });
+
+  it('should charge/transfer penalty', async function () {
+
+    let gatewayInitialBalance = await baseToken.balanceOf(gateway.address);
+    let stakerInitialBalance = await baseToken.balanceOf(stakeMessage.staker);
+
+    let penalty = new BN(150);
+
+    await baseToken.approve(gateway.address, penalty, {from: stakeMessage.staker});
+    await gateway.setOutboxStatus(
+      stakeMessage.messageHash,
+      MessageStatusEnum.Declared,
+    );
+
+    await gateway.revertStake(
+      stakeMessage.messageHash,
+      {from: stakeMessage.staker},
+    );
+
+    let gatewayFinalBalance = await baseToken.balanceOf(gateway.address);
+    let stakerFinalBalance = await baseToken.balanceOf(stakeMessage.staker);
+
+    assert.strictEqual(
+      gatewayFinalBalance.eq(gatewayInitialBalance.add(penalty)),
+      true,
+      'Penalty must be transferred to gateway',
+    );
+
+    assert.strictEqual(
+      stakerFinalBalance.eq(stakerInitialBalance.sub(penalty)),
+      true,
+      'Penalty must be transferred from staker',
+    );
+  });
+
+  it('should fail for zero message hash', async function () {
+
+    let penalty = new BN(150);
+
+    await baseToken.approve(gateway.address, penalty, {from: stakeMessage.staker});
+    await gateway.setOutboxStatus(
+      stakeMessage.messageHash,
+      MessageStatusEnum.Declared,
+    );
+
+    await Utils.expectRevert(
+      gateway.revertStake(
+        Utils.ZERO_BYTES32,
+        {from: stakeMessage.staker},
+      ),
+      'Message hash must not be zero.'
+      );
+
+  });
+
+  it('should fail if revocation is already declared', async function () {
+
+    let penalty = new BN(150);
+
+    await baseToken.approve(gateway.address, penalty, {from: stakeMessage.staker});
+    await gateway.setOutboxStatus(
+      stakeMessage.messageHash,
+      MessageStatusEnum.Declared,
+    );
+
+    await gateway.revertStake(
+      stakeMessage.messageHash,
+      {from: stakeMessage.staker},
+    );
+
+    await Utils.expectRevert(
+      gateway.revertStake(
+        stakeMessage.messageHash,
+        {from: stakeMessage.staker},
+      ),
+      'Message on source must be Declared.'
+    )
+  });
+
+  it('should fail for undeclared message', async function () {
```

Comment:
Is there a test that asserts the behavior in case there is no outbox state set at all for a message? Should there be one?

Reply:
wouldn't this case be similar to `Undeclared` status, because Message status is an enum with default value as zero of type uint8? 
If outbox doesn't have message status it will return default value which is undeclared (zero).  

What do you think?

Reply:
True 👍